File: packages/server/src/sync/sync.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import type {
  PullResponseDto
} from './dto/pull-response.dto';
import type {
  PushConflict,
  PushResponseDto,
  SyncOpDto
} from './dto/push-request.dto';
import { MetricsService } from '../metrics/metrics.service';
import type {
  CheckinPayload,
  ExistingCheckinRecord,
  ExistingHabitRecord,
  HabitPayload,
  ParentHabitRecord,
  TxClient
} from './sync.types';
import {
  buildCursorClause,
  calculateNextCursor,
  normalizeCustomDays,
  normalizeDate,
  normalizeReminderEnabled,
  normalizeReminderTime,
  normalizeSortOrder,
  normalizeTags,
  parseCursor,
  serializeCheckin,
  serializeHabit,
  serializeTombstone
} from './sync.utils';
import { HABIT_FREQUENCIES, HabitFrequency, HabitSchedule, normalizeSchedule, scheduleFromLegacy } from '@habbit-runner/shared';

@Injectable()
export class SyncService {
  private readonly logger = new Logger(SyncService.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly metrics: MetricsService
  ) {}

  async pull(userId: string, since?: string, traceId?: string): Promise<PullResponseDto> {
    const pullStart = Date.now();
    try {
      const cursor = parseCursor(since);
      const updatedFilter = cursor
        ? { AND: [buildCursorClause(cursor, 'updatedAt')] }
        : undefined;
      const deletedFilter = cursor
        ? { AND: [buildCursorClause(cursor, 'deletedAt')] }
        : undefined;

      const client = await this.prisma.getClient();
      const habits = await client.habit.findMany({
        where: {
          userId,
          ...updatedFilter
        },
        orderBy: [
          { updatedAt: 'asc' },
          { id: 'asc' }
        ],
        take: 200
      });

      const checkins = await client.checkin.findMany({
        where: {
          userId,
          ...updatedFilter
        },
        orderBy: [
          { updatedAt: 'asc' },
          { id: 'asc' }
        ],
        take: 200
      });

      const tombstones = await client.tombstone.findMany({
        where: {
          userId,
          ...deletedFilter
        },
        orderBy: [
          { deletedAt: 'asc' },
          { id: 'asc' }
        ],
        take: 200
      });

      const cursorCandidates = [
        ...habits.map((h: { updatedAt: Date; id: string }) => ({ updatedAt: h.updatedAt, id: h.id })),
        ...checkins.map((c: { updatedAt: Date; id: string }) => ({ updatedAt: c.updatedAt, id: c.id })),
        ...tombstones.map((t: { deletedAt: Date; id: string }) => ({ updatedAt: t.deletedAt, id: t.id }))
      ];

      const nextCursor = calculateNextCursor(cursorCandidates);
      const serverTime = new Date().toISOString();

      this.metrics.recordPull(
        Date.now() - pullStart,
        habits.length + checkins.length + tombstones.length
      );

      return {
        habits: habits.map(serializeHabit),
        checkins: checkins.map(serializeCheckin),
        tombstones: tombstones.map(serializeTombstone),
        nextCursor,
        serverTime
      };
    } catch (error) {
      this.metrics.recordError();
      this.logger.error(
        `sync pull failed userId=${userId} traceId=${traceId ?? 'n/a'}`,
        error instanceof Error ? error.stack : String(error)
      );
      throw error;
    }
  }

  async push(userId: string, ops: SyncOpDto[], traceId?: string): Promise<PushResponseDto> {
    const pushStart = Date.now();
    const applied: string[] = [];
    const conflicts: PushConflict[] = [];
    const serverTime = new Date().toISOString();

    try {
      const client = await this.prisma.getClient();
      await client.$transaction(async (tx) => {
        const txClient = tx as unknown as TxClient;
        for (const op of ops) {
          if (!op.id) {continue;}

          const deduplicated = await this.tryCreateLog(txClient, op.id);
          if (!deduplicated) {continue;}

          if (op.entity === 'habit') {
            await this.applyHabitOp(txClient, userId, op, applied, conflicts);
          } else if (op.entity === 'checkin') {
            await this.applyCheckinOp(txClient, userId, op, applied, conflicts);
          }
        }
      });
    } catch (error) {
      this.metrics.recordError();
      this.logger.error(
        `sync push failed userId=${userId} traceId=${traceId ?? 'n/a'}`,
        error instanceof Error ? error.stack : String(error)
      );
      throw error;
    }

    this.metrics.recordPush(Date.now() - pushStart, conflicts.length, ops.length);
    return { applied, conflicts, serverTime };
  }

  private async applyHabitOp(
    tx: TxClient,
    userId: string,
    op: SyncOpDto,
    applied: string[],
    conflicts: PushConflict[]
  ) {
    const payload = op.payload as unknown as HabitPayload;
    if (!payload?.id) {return;}
    const timestamp = normalizeDate(payload.updatedAt);

    if (op.type === 'delete') {
      await this.deleteHabit(tx, userId, payload);
      applied.push(op.id);
      return;
    }

    const existing = await tx.habit.findUnique({ where: { id: payload.id } }) as ExistingHabitRecord | null;
    if (this.hasHabitOwnershipConflict(existing, userId)) {
      conflicts.push({
        opId: op.id,
        reason: 'habit belongs to another user'
      });
      return;
    }
    if (this.hasNewerHabitConflict(existing, timestamp, op.id, conflicts)) {return;}

    await this.upsertHabit(tx, userId, payload, existing, timestamp);

    applied.push(op.id);
  }

  private async deleteHabit(tx: TxClient, userId: string, payload: HabitPayload): Promise<void> {
    await tx.tombstone.create({
      data: {
        userId,
        entity: 'habit',
        entityId: payload.id,
        version: payload.version ?? 1
      }
    });
    await tx.checkin.deleteMany({ where: { habitId: payload.id, userId } });
    await tx.habit.deleteMany({ where: { id: payload.id, userId } });
  }

  private hasHabitOwnershipConflict(existing: ExistingHabitRecord | null, userId: string): boolean {
    return Boolean(existing && existing.userId !== userId);
  }

  private hasNewerHabitConflict(
    existing: ExistingHabitRecord | null,
    timestamp: Date,
    opId: string,
    conflicts: PushConflict[]
  ): boolean {
    if (!existing || new Date(existing.updatedAt).getTime() <= timestamp.getTime()) {
      return false;
    }

    conflicts.push({
      opId,
      reason: 'server already has newer habit',
      serverValue: {
        version: existing.version,
        updatedAt: existing.updatedAt
      }
    });
    return true;
  }

  private async upsertHabit(
    tx: TxClient,
    userId: string,
    payload: HabitPayload,
    existing: ExistingHabitRecord | null,
    timestamp: Date
  ): Promise<void> {
    const writeValues = this.buildHabitWriteValues(payload, existing);
    await tx.habit.upsert({
      where: { id: payload.id },
      create: {
        id: payload.id,
        userId,
        name: payload.name,
        description: payload.description ?? null,
        color: payload.color,
        icon: payload.icon,
        frequency: payload.frequency,
        customDays: writeValues.customDays as never,
        schedule: writeValues.schedule as never,
        targetStreak: payload.targetStreak,
        dailyTarget: writeValues.dailyTarget,
        tags: writeValues.tags as never,
        archived: payload.archived ?? false,
        createdAt: normalizeDate(payload.createdAt),
        sortOrder: writeValues.sortOrder,
        reminderTime: writeValues.reminderTime,
        reminderEnabled: writeValues.reminderEnabled,
        difficulty: writeValues.difficulty,
        type: writeValues.type,
        updatedAt: timestamp,
        version: writeValues.nextVersion
      },
      update: {
        name: payload.name,
        description: payload.description ?? null,
        color: payload.color,
        icon: payload.icon,
        frequency: payload.frequency,
        customDays: writeValues.customDays as never,
        schedule: writeValues.schedule as never,
        targetStreak: payload.targetStreak,
        dailyTarget: writeValues.dailyTarget,
        tags: writeValues.tags as never,
        archived: payload.archived ?? false,
        sortOrder: writeValues.sortOrder,
        reminderTime: writeValues.reminderTime,
        reminderEnabled: writeValues.reminderEnabled,
        difficulty: writeValues.difficulty,
        type: writeValues.type,
        updatedAt: timestamp,
        version: writeValues.nextVersion
      }
    });
  }

  private buildHabitWriteValues(
    payload: HabitPayload,
    existing: ExistingHabitRecord | null
  ): {
    nextVersion: number;
    sortOrder: bigint;
    dailyTarget: number;
    reminderTime: string | null;
    reminderEnabled: boolean;
    tags: unknown;
    customDays: number[] | undefined;
    schedule: HabitSchedule;
    difficulty: number;
    type: string;
  } {
    return {
      nextVersion: this.resolveHabitVersion(existing, payload.version),
      sortOrder: this.resolveHabitSortOrder(existing, payload.sortOrder),
      dailyTarget: this.resolveHabitDailyTarget(existing, payload.dailyTarget),
      reminderTime: this.resolveHabitReminderTime(existing, payload.reminderTime),
      reminderEnabled: this.resolveHabitReminderEnabled(existing, payload.reminderEnabled),
      tags: normalizeTags(payload.tags),
      customDays: normalizeCustomDays(payload.customDays),
      schedule: this.resolveHabitSchedule(existing, payload),
      difficulty: this.resolveHabitDifficulty(existing, payload.difficulty),
      type: this.resolveHabitType(existing, payload.type)
    };
  }

  private resolveHabitVersion(existing: ExistingHabitRecord | null, payloadVersion?: number): number {
    return Math.max(existing?.version ?? 0, payloadVersion ?? 0) + 1;
  }

  private resolveHabitSortOrder(existing: ExistingHabitRecord | null, payloadSortOrder?: number): bigint {
    const normalizedSortOrder = normalizeSortOrder(payloadSortOrder);
    if (typeof normalizedSortOrder === 'number') {
      return BigInt(normalizedSortOrder);
    }

    const existingSortOrder = existing?.sortOrder;
    if (typeof existingSortOrder === 'bigint') {
      return existingSortOrder;
    }
    if (typeof existingSortOrder === 'number') {
      return BigInt(Math.trunc(existingSortOrder));
    }

    return 0n;
  }

  private resolveHabitDailyTarget(existing: ExistingHabitRecord | null, payloadDailyTarget?: number): number {
    if (typeof payloadDailyTarget === 'number' && Number.isFinite(payloadDailyTarget)) {
      return Math.max(1, Math.trunc(payloadDailyTarget));
    }
    return Math.max(1, existing?.dailyTarget ?? 1);
  }

  private resolveHabitReminderTime(
    existing: ExistingHabitRecord | null,
    payloadReminderTime?: string | null
  ): string | null {
    return normalizeReminderTime(payloadReminderTime ?? existing?.reminderTime);
  }

  private resolveHabitReminderEnabled(
    existing: ExistingHabitRecord | null,
    payloadReminderEnabled?: boolean
  ): boolean {
    return normalizeReminderEnabled(payloadReminderEnabled, existing?.reminderEnabled ?? true);
  }

  private resolveHabitSchedule(
    existing: ExistingHabitRecord | null,
    payload: HabitPayload
  ): HabitSchedule {
    const normalized = normalizeSchedule(payload.schedule);
    if (normalized) {
      return normalized;
    }
    if (existing?.schedule) {
      const inherited = normalizeSchedule(existing.schedule);
      if (inherited) {
        return inherited;
      }
    }
    const frequency = this.normalizeHabitFrequency(payload.frequency);
    return scheduleFromLegacy(frequency, normalizeCustomDays(payload.customDays));
  }

  private normalizeHabitFrequency(value: unknown): HabitFrequency {
    if (typeof value !== 'string') {
      return 'daily';
    }
    if ((HABIT_FREQUENCIES as readonly string[]).includes(value)) {
      return value as HabitFrequency;
    }
    return 'daily';
  }

  private async applyCheckinOp(
    tx: TxClient,
    userId: string,
    op: SyncOpDto,
    applied: string[],
    conflicts: PushConflict[]
  ) {
    const payload = op.payload as unknown as CheckinPayload;
    if (!payload?.habitId || !payload.date) {
      return;
    }
    const timestamp = normalizeDate(payload.updatedAt);
    const date = new Date(payload.date);
    const parentHabit = await this.findParentHabit(tx, payload.habitId);
    if (!this.canApplyCheckinForUser(parentHabit, userId, op.id, conflicts)) {
      return;
    }

    if (op.type === 'delete') {
      await this.deleteCheckin(tx, userId, payload, date);
      applied.push(op.id);
      return;
    }

    const existing = await tx.checkin.findFirst({
      where: { habitId: payload.habitId, date, userId }
    }) as ExistingCheckinRecord | null;
    if (this.hasNewerCheckinConflict(existing, timestamp, op.id, conflicts)) {
      return;
    }

    const nextVersion =
      Math.max(existing?.version ?? 0, payload.version ?? 0) + 1;
    const normalizedCount = Math.max(1, Math.trunc(payload.count ?? 1));

    await tx.checkin.upsert({
      where: {
        habit_date_unique: {
          habitId: payload.habitId,
          date
        }
      },
      create: {
        habitId: payload.habitId,
        userId,
        date,
        done: payload.done,
        count: normalizedCount,
        updatedAt: timestamp,
        version: nextVersion
      },
      update: {
        done: payload.done,
        count: normalizedCount,
        updatedAt: timestamp,
        version: nextVersion
      }
    });

    applied.push(op.id);
  }

  private async findParentHabit(tx: TxClient, habitId: string): Promise<ParentHabitRecord | null> {
    return tx.habit.findUnique({
      where: { id: habitId },
      select: { userId: true }
    }) as Promise<ParentHabitRecord | null>;
  }

  private canApplyCheckinForUser(
    parentHabit: ParentHabitRecord | null,
    userId: string,
    opId: string,
    conflicts: PushConflict[]
  ): boolean {
    if (parentHabit && parentHabit.userId === userId) {
      return true;
    }

    conflicts.push({
      opId,
      reason: 'checkin habit belongs to another user'
    });
    return false;
  }

  private async deleteCheckin(
    tx: TxClient,
    userId: string,
    payload: CheckinPayload,
    date: Date
  ): Promise<void> {
    await tx.tombstone.create({
      data: {
        userId,
        entity: 'checkin',
        entityId: payload.id ?? `${payload.habitId}:${payload.date}`,
        version: payload.version ?? 1
      }
    });
    await tx.checkin.deleteMany({
      where: { habitId: payload.habitId, date, userId }
    });
  }

  private hasNewerCheckinConflict(
    existing: ExistingCheckinRecord | null,
    timestamp: Date,
    opId: string,
    conflicts: PushConflict[]
  ): boolean {
    if (!existing || new Date(existing.updatedAt).getTime() <= timestamp.getTime()) {
      return false;
    }

    conflicts.push({
      opId,
      reason: 'server already has newer checkin',
      serverValue: {
        version: existing.version,
        updatedAt: existing.updatedAt
      }
    });
    return true;
  }

  private resolveHabitDifficulty(existing: ExistingHabitRecord | null, payloadValue?: number): number {
    if (typeof payloadValue === 'number') {
      return Math.max(1, Math.min(5, Math.trunc(payloadValue)));
    }
    return existing?.difficulty ?? 1;
  }

  private resolveHabitType(existing: ExistingHabitRecord | null, payloadValue?: string): string {
    if (typeof payloadValue === 'string') {
      return payloadValue === 'negative' ? 'negative' : 'positive';
    }
    return existing?.type ?? 'positive';
  }

  private async tryCreateLog(
    tx: TxClient,
    opId: string
  ): Promise<boolean> {
    const result = await tx.syncOpLog.createMany({
      data: [{ opId }],
      skipDuplicates: true
    });
    return result.count > 0;
  }
}

File: packages/server-java/src/main/java/com/habittracker/sync/SyncService.java
package com.habittracker.sync;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.habittracker.model.CheckinEntity;
import com.habittracker.model.HabitEntity;
import com.habittracker.model.SyncOpLogEntity;
import com.habittracker.model.TombstoneEntity;
import io.quarkus.hibernate.orm.panache.Panache;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;

import java.math.BigInteger;
import java.time.Instant;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@ApplicationScoped
public class SyncService {
  private static final DateTimeFormatter ISO = DateTimeFormatter.ISO_OFFSET_DATE_TIME;
  private static final TypeReference<Map<String, Object>> MAP_TYPE = new TypeReference<>() {
  };

  final ObjectMapper objectMapper;

  public SyncService(ObjectMapper objectMapper) {
    this.objectMapper = objectMapper;
  }

  public SyncDtos.PullResponseDto pull(String userId, String since) {
    var cursor = parseCursor(since);
    var habits = HabitEntity.<HabitEntity>find("userId", userId).list();
    var checkins = CheckinEntity.<CheckinEntity>find("userId", userId).list();
    var tombstones = TombstoneEntity.<TombstoneEntity>find("userId", userId).list();

    var filteredHabits = habits.stream()
        .filter(h -> cursor == null || isAfterCursor(h.updatedAt, h.id, cursor))
        .sorted(Comparator.comparing((HabitEntity h) -> h.updatedAt).thenComparing(h -> h.id))
        .limit(200)
        .toList();

    var filteredCheckins = checkins.stream()
        .filter(c -> cursor == null || isAfterCursor(c.updatedAt, c.id, cursor))
        .sorted(Comparator.comparing((CheckinEntity c) -> c.updatedAt).thenComparing(c -> c.id))
        .limit(200)
        .toList();

    var filteredTombstones = tombstones.stream()
        .filter(t -> cursor == null || isAfterCursor(t.deletedAt, t.id, cursor))
        .sorted(Comparator.comparing((TombstoneEntity t) -> t.deletedAt).thenComparing(t -> t.id))
        .limit(200)
        .toList();

    var candidates = new ArrayList<CursorRow>();
    filteredHabits.forEach(h -> candidates.add(new CursorRow(h.updatedAt, h.id)));
    filteredCheckins.forEach(c -> candidates.add(new CursorRow(c.updatedAt, c.id)));
    filteredTombstones.forEach(t -> candidates.add(new CursorRow(t.deletedAt, t.id)));

    var nextCursor = calculateNextCursor(candidates);

    return new SyncDtos.PullResponseDto(
        filteredHabits.stream().map(this::serializeHabit).toList(),
        filteredCheckins.stream().map(this::serializeCheckin).toList(),
        filteredTombstones.stream().map(this::serializeTombstone).toList(),
        nextCursor,
        toSyncIso(Instant.now())
    );
  }

  @Transactional
  public SyncDtos.PushResponseDto push(String userId, List<SyncDtos.SyncOpDto> ops) {
    var applied = new ArrayList<String>();
    var conflicts = new ArrayList<SyncDtos.PushConflict>();

    for (var op : ops) {
      if (op.id() == null || op.id().isBlank()) {
        continue;
      }
      if (!tryCreateLog(op.id())) {
        continue;
      }
      if ("habit".equals(op.entity())) {
        applyHabitOp(userId, op, applied, conflicts);
      } else if ("checkin".equals(op.entity())) {
        applyCheckinOp(userId, op, applied, conflicts);
      }
    }

    return new SyncDtos.PushResponseDto(applied, conflicts, toSyncIso(Instant.now()));
  }

  private void applyHabitOp(String userId, SyncDtos.SyncOpDto op, List<String> applied, List<SyncDtos.PushConflict> conflicts) {
    var payload = toMap(op.payload());
    var habitId = asString(payload.get("id"));
    if (habitId == null) {
      return;
    }

    if ("delete".equals(op.type())) {
      deleteHabit(userId, habitId, payload);
      applied.add(op.id());
      return;
    }

    var existing = (HabitEntity) HabitEntity.findById(habitId);
    if (existing != null && !userId.equals(existing.userId)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "habit belongs to another user", null));
      return;
    }

    var clientUpdated = normalizeInstant(asString(payload.get("updatedAt")));
    if (existing != null && existing.updatedAt.isAfter(clientUpdated)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "server already has newer habit", Map.of(
          "version", existing.version,
          "updatedAt", toSyncIso(existing.updatedAt)
      )));
      return;
    }

    if (existing == null) {
      existing = new HabitEntity();
      existing.id = habitId;
      existing.userId = userId;
      existing.createdAt = normalizeInstant(asString(payload.get("createdAt")));
      if (existing.createdAt == null) {
        existing.createdAt = Instant.now();
      }
      existing.persist();
    }

    existing.name = asString(payload.get("name"));
    existing.description = nullableString(payload.get("description"));
    existing.color = asString(payload.get("color"));
    existing.icon = asString(payload.get("icon"));
    existing.frequency = asString(payload.get("frequency"));
    existing.customDays = jsonOrNull(payload.get("customDays"));
    existing.schedule = jsonOrNull(payload.get("schedule"));
    existing.targetStreak = asInt(payload.get("targetStreak"), 1);
    existing.dailyTarget = Math.max(1, asInt(payload.get("dailyTarget"), existing.dailyTarget > 0 ? existing.dailyTarget : 1));
    existing.tags = jsonOrNull(payload.get("tags"));
    existing.archived = asBoolean(payload.get("archived"), false);
    existing.sortOrder = BigInteger.valueOf(asInt(payload.get("sortOrder"), 0));
    existing.reminderTime = nullableString(payload.get("reminderTime"));
    existing.reminderEnabled = asBoolean(payload.get("reminderEnabled"), true);
    existing.type = normalizeType(asString(payload.get("type")));
    existing.freezeDays = jsonOrDefault(payload.get("freezeDays"), "[]");
    existing.version = Math.max(existing.version, asInt(payload.get("version"), 0)) + 1;
    existing.updatedAt = nextSyncDate(clientUpdated, existing.updatedAt);

    applied.add(op.id());
  }

  private void applyCheckinOp(String userId, SyncDtos.SyncOpDto op, List<String> applied, List<SyncDtos.PushConflict> conflicts) {
    var payload = toMap(op.payload());
    var habitId = asString(payload.get("habitId"));
    var dateString = asString(payload.get("date"));
    if (habitId == null || dateString == null) {
      return;
    }

    var parent = (HabitEntity) HabitEntity.findById(habitId);
    if (parent == null || !userId.equals(parent.userId)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "checkin habit belongs to another user", null));
      return;
    }

    var date = toLocalDate(dateString);
    var existing = CheckinEntity.<CheckinEntity>find("habitId = ?1 and date = ?2 and userId = ?3", habitId, date, userId).firstResult();
    var clientUpdated = normalizeInstant(asString(payload.get("updatedAt")));

    if ("delete".equals(op.type())) {
      deleteCheckin(userId, habitId, date, payload, existing);
      applied.add(op.id());
      return;
    }

    if (existing != null && existing.updatedAt.isAfter(clientUpdated)) {
      conflicts.add(new SyncDtos.PushConflict(op.id(), "server already has newer checkin", Map.of(
          "version", existing.version,
          "updatedAt", toSyncIso(existing.updatedAt)
      )));
      return;
    }

    if (existing == null) {
      existing = new CheckinEntity();
      existing.habitId = habitId;
      existing.userId = userId;
      existing.date = date;
      existing.persist();
    }

    existing.done = asBoolean(payload.get("done"), false);
    existing.count = Math.max(1, asInt(payload.get("count"), 1));
    existing.version = Math.max(existing.version, asInt(payload.get("version"), 0)) + 1;
    existing.updatedAt = nextSyncDate(clientUpdated, existing.updatedAt);

    applied.add(op.id());
  }

  private void deleteHabit(String userId, String habitId, Map<String, Object> payload) {
    var tombstone = new TombstoneEntity();
    tombstone.userId = userId;
    tombstone.entity = "habit";
    tombstone.entityId = habitId;
    tombstone.version = asInt(payload.get("version"), 1);
    tombstone.deletedAt = nextSyncDate(normalizeInstant(asString(payload.get("updatedAt"))), null);
    tombstone.persist();

    Panache.executeUpdate("delete from CheckinEntity where habitId = ?1 and userId = ?2", habitId, userId);
    Panache.executeUpdate("delete from HabitEntity where id = ?1 and userId = ?2", habitId, userId);
  }

  private void deleteCheckin(String userId, String habitId, LocalDate date, Map<String, Object> payload, CheckinEntity existing) {
    var tombstone = new TombstoneEntity();
    tombstone.userId = userId;
    tombstone.entity = "checkin";
    var payloadId = asString(payload.get("id"));
    tombstone.entityId = payloadId != null ? payloadId : (existing != null ? existing.id : habitId + ":" + date);
    tombstone.version = asInt(payload.get("version"), 1);
    tombstone.deletedAt = nextSyncDate(normalizeInstant(asString(payload.get("updatedAt"))), null);
    tombstone.persist();

    Panache.executeUpdate("delete from CheckinEntity where habitId = ?1 and userId = ?2 and date = ?3", habitId, userId, date);
  }

  private boolean tryCreateLog(String opId) {
    var existing = SyncOpLogEntity.findById(opId);
    if (existing != null) {
      return false;
    }
    var log = new SyncOpLogEntity();
    log.opId = opId;
    log.persist();
    return true;
  }

  private SyncDtos.HabitDto serializeHabit(HabitEntity habit) {
    return new SyncDtos.HabitDto(
        habit.id,
        habit.name,
        habit.description == null ? "" : habit.description,
        habit.color,
        habit.icon,
        habit.frequency,
        parseJsonOrNull(habit.customDays),
        parseJsonOrNull(habit.schedule),
        habit.targetStreak,
        habit.dailyTarget,
        parseJsonOrEmptyList(habit.tags),
        habit.archived,
        toSyncIso(habit.createdAt),
        toSyncIso(habit.updatedAt),
        habit.version,
        habit.sortOrder == null ? 0 : habit.sortOrder.intValue(),
        habit.reminderTime,
        habit.reminderEnabled,
        habit.type,
        parseJsonOrEmptyList(habit.freezeDays)
    );
  }

  private SyncDtos.CheckinDto serializeCheckin(CheckinEntity checkin) {
    return new SyncDtos.CheckinDto(
        checkin.id,
        checkin.habitId,
        checkin.date.toString(),
        checkin.done,
        checkin.count,
        toSyncIso(checkin.updatedAt),
        checkin.version
    );
  }

  private SyncDtos.TombstoneDto serializeTombstone(TombstoneEntity tombstone) {
    return new SyncDtos.TombstoneDto(
        tombstone.id,
        tombstone.entity,
        tombstone.entityId,
        toSyncIso(tombstone.deletedAt),
        tombstone.version
    );
  }

  private boolean isAfterCursor(Instant time, String id, CursorRow cursor) {
    if (time == null) {
      return false;
    }
    if (time.isAfter(cursor.updatedAt())) {
      return true;
    }
    if (time.isBefore(cursor.updatedAt())) {
      return false;
    }
    return id.compareTo(cursor.id()) > 0;
  }

  private CursorRow parseCursor(String raw) {
    if (raw == null || raw.isBlank()) {
      return null;
    }
    try {
      var data = objectMapper.readValue(raw, MAP_TYPE);
      var updatedAt = normalizeInstant(asString(data.get("updatedAt")));
      var id = asString(data.get("id"));
      if (updatedAt == null || id == null) {
        return null;
      }
      return new CursorRow(updatedAt, id);
    } catch (Exception ex) {
      return null;
    }
  }

  private String calculateNextCursor(List<CursorRow> rows) {
    if (rows.isEmpty()) {
      return null;
    }
    var latest = rows.stream().max((a, b) -> {
      var cmp = a.updatedAt().compareTo(b.updatedAt());
      if (cmp != 0) {
        return cmp;
      }
      return a.id().compareTo(b.id());
    }).orElse(null);

    if (latest == null) {
      return null;
    }

    try {
      return objectMapper.writeValueAsString(Map.of(
          "updatedAt", toSyncIso(latest.updatedAt()),
          "id", latest.id()
      ));
    } catch (Exception ex) {
      return null;
    }
  }

  private Instant normalizeInstant(String value) {
    if (value == null || value.isBlank()) {
      return Instant.now();
    }
    try {
      return Instant.parse(value);
    } catch (Exception ex) {
      return Instant.now();
    }
  }

  private LocalDate toLocalDate(String value) {
    try {
      if (value.length() >= 10) {
        return LocalDate.parse(value.substring(0, 10));
      }
    } catch (Exception ignored) {
    }
    return LocalDate.now(ZoneOffset.UTC);
  }

  private Instant nextSyncDate(Instant... values) {
    var max = Instant.now();
    for (var value : values) {
      if (value != null && value.isAfter(max)) {
        max = value;
      }
    }
    return max.plusSeconds(1);
  }

  private String toSyncIso(Instant instant) {
    return ISO.format(OffsetDateTime.ofInstant(instant, ZoneOffset.UTC));
  }

  private Map<String, Object> toMap(Map<String, Object> payload) {
    return payload == null ? new HashMap<>() : payload;
  }

  private String asString(Object value) {
    if (value == null) {
      return null;
    }
    var text = String.valueOf(value);
    return text.isBlank() ? null : text;
  }

  private String nullableString(Object value) {
    if (value == null) {
      return null;
    }
    return String.valueOf(value);
  }

  private int asInt(Object value, int fallback) {
    if (value instanceof Number number) {
      return number.intValue();
    }
    if (value instanceof String text) {
      try {
        return Integer.parseInt(text);
      } catch (Exception ignored) {
      }
    }
    return fallback;
  }

  private boolean asBoolean(Object value, boolean fallback) {
    if (value instanceof Boolean bool) {
      return bool;
    }
    if (value instanceof String text) {
      return "true".equalsIgnoreCase(text) || "1".equals(text);
    }
    return fallback;
  }

  private String normalizeType(String type) {
    return "negative".equals(type) ? "negative" : "positive";
  }

  private String jsonOrNull(Object value) {
    if (value == null) {
      return null;
    }
    try {
      return objectMapper.writeValueAsString(value);
    } catch (Exception ex) {
      return null;
    }
  }

  private String jsonOrDefault(Object value, String fallbackJson) {
    var result = jsonOrNull(value);
    return result == null ? fallbackJson : result;
  }

  private Object parseJsonOrNull(String json) {
    if (json == null || json.isBlank()) {
      return null;
    }
    try {
      return objectMapper.readValue(json, Object.class);
    } catch (Exception ex) {
      return null;
    }
  }

  private Object parseJsonOrEmptyList(String json) {
    var parsed = parseJsonOrNull(json);
    return parsed == null ? List.of() : parsed;
  }

  private record CursorRow(Instant updatedAt, String id) {
  }
}
